app/Services/Validation/ApplicationTimelineValidator.php

```diff
@@ -0,0 +1,420 @@
+
+        // Check to see if this application is related to a Job Poster with
+        // a bilingual requirement.
+        $langRequirement = $application->job_poster->language_requirement_id;
+        $langTestRequirement = LanguageRequirement::where('name', 'bilingual_intermediate')
+            ->orWhere('name', 'bilingual_advanced')
+            ->pluck('id');
+
+        $validator = Validator::make($application->toArray(), $this->basicInfoSimpleRules());
+
+        // Conditionally check for 'language_test_confirmed' if the
+        // closure returns true.
+        $validator->sometimes(
+            'language_test_confirmed',
+            'required|boolean|accepted',
+            function ($input) use ($langRequirement, $langTestRequirement) {
+                return in_array($langRequirement, $langTestRequirement->toArray());
+            }
+        );
+
+        return $validator;
+    }
+
+    /**
+     * Helper function to return completeness for the Basic Info step.
+     *
+     * @param JobApplication $application Job Application object.
+     *
+     * @return boolean
+     */
+    public function basicsComplete(JobApplication $application)
+    {
+        $validator = $this->basicsValidator($application);
+        return $validator->passes();
+    }
+
+    /* End Step 1 --------------------------------------------------------------------------------------------------- */
+
+    /* Step 2 ------------------------------------------------------------------------------------------------------- */
+
+    /**
+     * Validator instance for the Experience step.
+     *
+     * @param JobApplication $application Job Application object.
+     *
+     * @return \Illuminate\Contracts\Validation\Validator
+     */
+    public function experienceValidator(JobApplication $application)
+    {
+        $application->loadMissing('job_poster', 'job_poster.criteria');
+
+        $essentialCriteriaType = CriteriaType::where('name', 'essential')->first()->id;
+
+        $jobCriteria = $application->job_poster->criteria;
+        $requiredCriteria = $jobCriteria->filter(function ($criterion) use ($essentialCriteriaType) {
+            return $criterion->criteria_type_id === $essentialCriteriaType;
+        });
+
+        $experienceSkillRules = array();
+
+        foreach ($requiredCriteria as $criterion) {
+            // Validate that every essential skill has a corresponding experience.
+            $experienceSkillRules[] = [
+                'skill_id' => new ContainsArrayWithAttributeRule('skill_id', $criterion->skill_id)
+            ];
+        }
+
+        // Get all Experiences belonging to the current application, that are assigned to required Criteria.
+        $experiences = ExperienceSkill::whereHasMorph(
+            'experience',
+            '*',
+            function (Builder $query) use ($application): void {
+                $query->where([
+                    ['experienceable_type', 'application'],
```

Comment:
The experienceable_type depends on the $application status. If `$application->isDraft()`, experienceable_type must be 'applicant'. (And since we mainly need to validate _before_ submitting, this will usually be the case.)

Reply:
And of course, it type is 'applicant', id will need to be `$application->applicant_id`.